### ISSUE TITLE
The ability to compose tokens

### ISSUE THREAD

**Issue by benlesh** [zenparsing/es-cancel-token#6]
The idea here is that you can combine tokens so one token's cancellation can be linked to another. This would make CancelToken more useful for composing chains of async, as found in RxJS or other observable libraries. For example:

``` js
let token1 = new CancelToken();
let token2 = new CancelToken();

token1.add(token2);

token1.cancel(); // both 1 and 2 are cancelled
```

``` js
let token1 = new CancelToken();
let token2 = new CancelToken();

token1.add(token2);

token2.cancel(); // only 2 is cancelled
```

``` js
let token1 = new CancelToken();
let token2 = new CancelToken();

token1.add(token2);
token2.add(token1);

token2.cancel(); // both 1 and 2 are cancelled
```

``` js
let token1 = new CancelToken();
let token2 = new CancelToken();
let token3 = new CancelToken();

token1.add(token2);
token2.add(token2);

token1.cancel(); // both 1 and 2 are cancelled, not 3
```

``` js
let token1 = new CancelToken();
let token2 = new CancelToken();
let token3 = new CancelToken();

token1.add(token2);
token2.add(token2);

token2.cancel(); // only 2 is cancelled
```

**Comment by benjamingr** [comment]
Note that this isn't useful only for observables - promise APIs, async iterators DOM and Node events are all potential beneficiaries of this. One way is easy (registering for a notification when all tokens are registered).

An argument against this would be that it's reasonably easy to achieve this behavior with an array or to implement it in userland.

**Comment by domenic** [comment]
Yes, this use case generally make sense to me. We should probably look to .NET for inspiration at least in breadth of API. I'd appreciate some small motivating real-worldish examples. (Preferably for application authors, since that's easier to explain, although of course library authors are super-important as well.)

Is a mutable API important, or is it possible to achieve the same thing with a (cancel token, cancel token) -> cancel token combinator? For example, if you posit a combinator `CancelToken.cancelBoth(t1, t2)` which returns a `cancel` function that cancels both `t1` and `t2`, I think I can rewrite all your above examples to work.

But I'm not sure I entirely understand the point of your examples, as they seem to be illustrating mechanism, not use case---for example the second example would behave the same if you removed the `.add()` call entirely.

**Comment by benjamingr** [comment]
Note that a C# cancellation token is a value type. Assuming that's too much work and ES value types are not even a thing anymore from what I understand - we should consider an immutable API where:

``` js
token1.add(token2)
```

Does not alter token1 but instead returns a new token. I'm even more in favor of a `Promise.all` like API where we do:

``` js
var token4 = CancelationToken.all(token1, token2, token3);
```

I agree we should see use cases first though. C# has aggregation built in for sources but not tokens themselves: https://msdn.microsoft.com/en-us/library/dd642265(v=vs.110).aspx

**Comment by domenic** [comment]
I was hoping not to make sources a first-class type, but it's possible that wanting combinators forces that to happen. Use cases would be helpful for judging whether it's possible to create combinators for those use cases without creating a new source type.

**Comment by bergus** [comment]
I've implemented a few combinators that I found useful in [bergus/creed:cancellation](//github.com/bergus/creed/tree/cancellation) (for the curious, the pretty unreadable and proprietary implementation - tailored to the library style - can be found in [CancelToken.js](//github.com/bergus/creed/blob/cancellation/src/CancelToken.js)). There are
- `p = CancelToken.pool()` - creates a Pool object
  - to which you can `p.add(token1, token2, …)`
  - whose `p.get()` token cancels with an array of reasons once all (>=1) of the added tokens are requested
- `r = CancelToken.race()` - creates a Race object
  - to which you can `r.add(token1, token2, …)`
  - whose `p.get()` token cancels with the reason of the first requested of the added tokens
- `r = CancelToken.reference()` - creates a Reference object
  - whose reference you can `r.set(token)` or `r.set(null)` (overwriting previously set ones)
  - whose `r.get()` token cancels with the reason of the current reference once that is requested

All the factory functions also take an argument to initialise the combinator without the `add`/`set` method.
Once the resulting token is cancelled, further `add`/`set` calls don't have any effect.
The resulting tokens are pretty special in that their `.requested` getter immediately (synchronously) determines the status, which means that their state is always correct regardless of asynchronous subscription to the tokens.

The `reference` was useful to implement the behaviour of the magic `await.cancel` token I proposed in domenic/cancelable-promise#19 / domenic/cancelable-promise#23. The `pool` is supposed to be useful for a `Task` implementation, especially with the dynamic addition of tokens (that's why I named it `pool` not `all`, maybe one should even be allowed to remove tokens). The `race` can be used to implement a `token1.concat(token2)` method that returns a new token which cancels with either cancellation.

`CancelToken`s themselves should be immutable (that's the reason why they don't have a `.cancel()` method either, only their sources do), the combinators I posed above might not.
